File: backend/app/services/return_service.py
"""Vehicle return service (Module 7).

Computes extra days, late fees, damage costs and the final payment owed (or to
be refunded) when a car is returned.
"""
from __future__ import annotations

from datetime import date, datetime
from typing import Any

from app.models import VehicleReturn
from app.models.enums import CarStatus, RentalStatus
from app.repositories import car_repo, rental_repo
from app.services import audit_service, car_service
from app.utils.errors import ConflictError, ValidationError


def _parse_datetime(value: Any) -> datetime:
    if isinstance(value, datetime):
        return value
    if not value:
        return datetime.now()
    text = str(value)
    try:
        return datetime.fromisoformat(text)
    except ValueError:
        return datetime.combine(date.fromisoformat(text[:10]), datetime.min.time())


def calculate_return(
    rental,
    *,
    return_date: datetime,
    damage_cost: float = 0.0,
    penalties: float = 0.0,
    late_fee_multiplier: float = 1.5,
) -> dict:
    """Calculate the financial outcome of a return.

    * Extra days are days returned after the agreed end date.
    * Late fee = extra_days * daily_price * multiplier.
    * Final payment = late fee + damage cost + penalties - deposit.
      A negative value means the company must refund the client.
    """
    agreed_end = rental.rental_end
    actual_end = return_date.date() if isinstance(return_date, datetime) else return_date

    extra_days = max((actual_end - agreed_end).days, 0)
    late_fee = round(extra_days * rental.daily_price * late_fee_multiplier, 2)
    final_payment = round(late_fee + damage_cost + penalties - rental.deposit, 2)

    return {
        "extra_days": extra_days,
        "late_fee": late_fee,
        "damage_cost": round(damage_cost, 2),
        "penalties": round(penalties, 2),
        "final_payment": final_payment,
    }


def create_return(rental_id: int, data: dict[str, Any], actor=None) -> VehicleReturn:
    rental = rental_repo.get_or_404(rental_id)
    if rental.vehicle_return is not None:
        raise ConflictError("Возврат для этой аренды уже оформлен")
    if rental.status == RentalStatus.CANCELLED:
        raise ValidationError("Невозможно оформить возврат для отменённой аренды")

    return_date = _parse_datetime(data.get("return_date"))
    damage_cost = float(data.get("damage_cost") or 0)
    penalties = float(data.get("penalties") or 0)

    calc = calculate_return(
        rental,
        return_date=return_date,
        damage_cost=damage_cost,
        penalties=penalties,
    )

    vehicle_return = VehicleReturn(
        rental_id=rental.id,
        return_date=return_date,
        mileage=int(data.get("mileage") or rental.car.mileage),
        fuel_level=int(data.get("fuel_level") or 100),
        damages=data.get("damages"),
        extra_days=calc["extra_days"],
        late_fee=calc["late_fee"],
        damage_cost=calc["damage_cost"],
        penalties=calc["penalties"],
        final_payment=calc["final_payment"],
    )
    rental.vehicle_return = vehicle_return
    rental.status = RentalStatus.COMPLETED

    # Update odometer and free the car.
    car = car_repo.get(rental.car_id)
    if car:
        car.mileage = vehicle_return.mileage
        car_service.set_status(car, CarStatus.AVAILABLE)

    rental_repo.update(rental, {})

    audit_service.record(
        "create_return",
        user_id=getattr(actor, "id", None),
        username=getattr(actor, "username", None),
        entity="rental",
        entity_id=rental.id,
        details=f"Возврат: итог {calc['final_payment']:.2f}",
    )
    return vehicle_return

File: backend-go/internal/service/return_service.go
package service

import (
	"time"

	"github.com/instaagrammeta/rentacar/backend-go/internal/apperr"
	"github.com/instaagrammeta/rentacar/backend-go/internal/models"
	"github.com/instaagrammeta/rentacar/backend-go/internal/pricing"
)

// ReturnCalc holds the financial outcome of a vehicle return.
type ReturnCalc struct {
	ExtraDays    int     `json:"extra_days"`
	LateFee      float64 `json:"late_fee"`
	DamageCost   float64 `json:"damage_cost"`
	Penalties    float64 `json:"penalties"`
	FinalPayment float64 `json:"final_payment"`
}

// ToMap renders the calculation as a JSON map.
func (c ReturnCalc) ToMap() map[string]interface{} {
	return map[string]interface{}{
		"extra_days":    c.ExtraDays,
		"late_fee":      c.LateFee,
		"damage_cost":   c.DamageCost,
		"penalties":     c.Penalties,
		"final_payment": c.FinalPayment,
	}
}

// CalculateReturn computes extra days, late fee and the final payment owed.
// A negative final payment means the company must refund the client.
func (s *Service) CalculateReturn(rental *models.Rental, returnDate time.Time, damageCost, penalties float64) ReturnCalc {
	agreedEnd := rental.RentalEnd
	actualEnd := returnDate

	extraDays := int(actualEnd.Sub(agreedEnd).Hours() / 24)
	if extraDays < 0 {
		extraDays = 0
	}
	lateFee := pricing.Round2(float64(extraDays) * rental.DailyPrice * s.Cfg.LateFeeMultiplier)
	finalPayment := pricing.Round2(lateFee + damageCost + penalties - rental.Deposit)

	return ReturnCalc{
		ExtraDays:    extraDays,
		LateFee:      lateFee,
		DamageCost:   pricing.Round2(damageCost),
		Penalties:    pricing.Round2(penalties),
		FinalPayment: finalPayment,
	}
}

// ReturnInput holds fields for creating a vehicle return.
type ReturnInput struct {
	ReturnDate string   `json:"return_date"`
	Mileage    *int     `json:"mileage"`
	FuelLevel  *int     `json:"fuel_level"`
	Damages    *string  `json:"damages"`
	DamageCost *float64 `json:"damage_cost"`
	Penalties  *float64 `json:"penalties"`
}

// CreateReturn records a vehicle return and completes the rental.
func (s *Service) CreateReturn(rentalID uint, in ReturnInput, actor *Actor) (*models.VehicleReturn, error) {
	rental, err := s.GetRental(rentalID)
	if err != nil {
		return nil, err
	}
	if rental.VehicleReturn != nil {
		return nil, apperr.Conflict("Возврат для этой аренды уже оформлен")
	}
	if rental.Status == models.RentalCancelled {
		return nil, apperr.Validation("Невозможно оформить возврат для отменённой аренды")
	}

	returnDate, _ := parseDateTime(in.ReturnDate)
	damageCost := derefFloat(in.DamageCost)
	penalties := derefFloat(in.Penalties)

	calc := s.CalculateReturn(rental, returnDate, damageCost, penalties)

	mileage := 0
	if rental.Car != nil {
		mileage = rental.Car.Mileage
	}
	if in.Mileage != nil {
		mileage = *in.Mileage
	}
	fuel := 100
	if in.FuelLevel != nil {
		fuel = *in.FuelLevel
	}

	vr := &models.VehicleReturn{
		RentalID:     rental.ID,
		ReturnDate:   returnDate,
		Mileage:      mileage,
		FuelLevel:    fuel,
		Damages:      in.Damages,
		ExtraDays:    calc.ExtraDays,
		LateFee:      calc.LateFee,
		DamageCost:   calc.DamageCost,
		Penalties:    calc.Penalties,
		FinalPayment: calc.FinalPayment,
	}
	if err := s.DB.Create(vr).Error; err != nil {
		return nil, err
	}

	s.DB.Model(&models.Rental{}).Where("id = ?", rental.ID).Update("status", models.RentalCompleted)

	// Update odometer and free the car.
	if rental.Car != nil {
		s.DB.Model(&models.Car{}).Where("id = ?", rental.CarID).
			Updates(map[string]interface{}{"mileage": mileage, "status": models.CarAvailable})
	}

	s.record(actor, "create_return", "rental", uintPtr(rental.ID),
		"Возврат: итог "+formatMoney(calc.FinalPayment))
	return vr, nil
}
